src/Data/Bifunctor/Linear.hs

```diff
@@ -0,0 +1,59 @@
+{-# LANGUAGE FunctionalDependencies #-}
+{-# LANGUAGE KindSignatures #-}
+{-# LANGUAGE LinearTypes #-}
+{-# LANGUAGE NoImplicitPrelude #-}
+{-# LANGUAGE TypeOperators #-}
+
+module Data.Bifunctor.Linear where
+
+import Prelude.Linear
+import Data.Void
+
+-- TODO: associators
+-- TODO: remove `swap` from Prelude
+
+-- | Bifunctors on the category of linear types
+--
+-- Laws:
+-- If 'bimap' is supplied, then
+-- * @'bimap' 'id' 'id' = 'id'@
+-- If 'first' and 'second' are supplied, then
+-- * @
+-- 'first' 'id' ≡ 'id'
+-- 'second' 'id' ≡ 'id'
+-- @
+-- If all are supplied, then
+-- * @'bimap' f g = 'first' f '.' 'second' g
+class Bifunctor p where
+  {-# MINIMAL bimap | (first , second) #-}
+  bimap :: (a ->. b) -> (c ->. d) -> a `p` c ->. b `p` d
```

Comment:
I wonder now if the final arrow here should actually be linear, comparing with the type

```haskell
dimap :: (s ->. a) -> (b ->. t) -> a `arr` b -> s `arr` t
```
in which the final arrow is non-linear.

Consistency feels like the correct choice, but I'm not sure if both should be linear or both non-linear.

Reply:
I think linear is the correct choice now. In particular, asking for 
```haskell
dimap :: (s ->. a) -> (b ->. t) -> a `arr` b ->. s `arr` t
```
would mean `rmap` witnesses that a `Profunctor` is a `Data.Functor` in its second argument, which feels more correct.

Reply:
Ah, `Bifunctor` has the `Data` vs `Control` issue again.

Profunctors are not endofunctors (they are functors to Set). So they don't have (quite) the same issue. `dimap` is supposed to be a non-linear arrow at the end in theory, and quite possibly, that would include more potential profunctors, which we may need (I'd say probably, even) to define `set` and `view` (these are not linear operations, but they still work when the data is not linear).

Looking concretely, we are considering `(,)` and `Either`, `Either` is not a Control bifunctor. So I'd say we can restrict our attention to Data Bifunctors for the time being. Hence `bimap :: (a ->. b) -> (c ->. d) -> a `p` c -> b `p` d`.

Reply:
I don't think this is the `Data` vs `Control` issue, since the `(a ->. b)` is not necessarily consumed linearly. Concretely, there's no issue making `Either` a `Bifunctor` with the `bimap` I gave above (as witnessed by the CI check passing).